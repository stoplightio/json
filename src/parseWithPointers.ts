import { DiagnosticSeverity, IDiagnostic, IParserASTResult, IParserResult, IRange } from '@stoplight/types';
import { JSONVisitor, Node, NodeType, ParseErrorCode, printParseErrorCode, visit } from 'jsonc-parser';
import { IJsonASTNode, IParseOptions } from './types';

export const parseWithPointers = <T = any>(
  value: string,
  options: IParseOptions = { disallowComments: true }
): IParserResult<T, IJsonASTNode, number[]> => {
  const diagnostics: IDiagnostic[] = [];
  const { ast, data, lineMap } = parseTree<T>(value, diagnostics, options);

  return {
    data,
    diagnostics,
    ast,
    lineMap,
  };
};

// based on source code of "https://github.com/Microsoft/node-jsonc-parser
/*---------------------------------------------------------------------------------------------
 *  Copyright (c) Microsoft Corporation. All rights reserved.
 *  Licensed under the MIT License. See License.txt in the project root for license information.
 *--------------------------------------------------------------------------------------------*/
export function parseTree<T>(
  text: string,
  errors: IDiagnostic[] = [],
  options: IParseOptions
): IParserASTResult<T, Node, number[]> {
  const lineMap = computeLineMap(text);
  let currentParent: IJsonASTNode = { type: 'array', offset: -1, length: -1, children: [], parent: void 0 }; // artificial root
  let currentParsedProperty: string | null = null;
  let currentParsedParent: any = [];
  const currentObjectKeys: string[] = [''];
  const previousParsedParents: any[] = [];

  function ensurePropertyComplete(endOffset: number) {
    if (currentParent.type === 'property') {
      currentParent.length = endOffset - currentParent.offset;
      currentParent = currentParent.parent!;
    }
  }

  function calculateRange(startLine: number, startCharacter: number, length: number): IRange {
    return {
      start: {
        line: startLine,
        character: startCharacter,
      },
      end: {
        line: startLine,
        character: startCharacter + length,
      },
    };
  }

  function onValue(valueNode: IJsonASTNode): IJsonASTNode {
    currentParent.children!.push(valueNode);
    return valueNode;
  }

  function onParsedValue(value: any) {
    if (Array.isArray(currentParsedParent)) {
      (currentParsedParent as any[]).push(value);
    } else if (currentParsedProperty) {
      currentParsedParent[currentParsedProperty] = value;
    }
  }

  function onParsedComplexBegin(value: any) {
    onParsedValue(value);
    previousParsedParents.push(currentParsedParent);
    currentParsedParent = value;
    currentParsedProperty = null;
  }

  function onParsedComplexEnd() {
    currentParsedParent = previousParsedParents.pop();
  }

  const visitor: JSONVisitor = {
    onObjectBegin: (offset, length, startLine, startCharacter: number) => {
      currentParent = onValue({
        type: 'object',
        offset,
        length: -1,
        parent: currentParent,
        children: [],
        range: calculateRange(startLine, startCharacter, length),
      });

      if (options.ignoreDuplicateKeys === false) {
        currentObjectKeys.length = 0;
      }

      onParsedComplexBegin({});
    },
    onObjectProperty: (name: string, offset: number, length: number, startLine: number, startCharacter: number) => {
      currentParent = onValue({ type: 'property', offset, length: -1, parent: currentParent, children: [] });
      currentParent.children!.push({ type: 'string', value: name, offset, length, parent: currentParent });

      if (options.ignoreDuplicateKeys === false) {
        if (currentObjectKeys.length === 0 || !currentObjectKeys.includes(name)) {
          currentObjectKeys.push(name);
        } else {
          errors.push({
            range: calculateRange(startLine, startCharacter, length),
            message: 'DuplicateKey',
            severity: DiagnosticSeverity.Error,
            code: 20, // 17 is the lowest safe value, but decided to bump it
          });
        }
      }

      currentParsedProperty = name;
    },
    onObjectEnd: (offset: number, length, startLine, startCharacter) => {
      currentParent.length = offset + length - currentParent.offset;
      if (currentParent.range) {
        // @ts-ignore, read only ;P
        currentParent.range.end.line = startLine;
        // @ts-ignore, read only ;P
        currentParent.range.end.character = startCharacter + length;
      }
      currentParent = currentParent.parent!;
      ensurePropertyComplete(offset + length);

      onParsedComplexEnd();
    },
    onArrayBegin: (offset, length, startLine, startCharacter) => {
      currentParent = onValue({
        type: 'array',
        offset,
        length: -1,
        parent: currentParent,
        children: [],
        range: calculateRange(startLine, startCharacter, length),
      });

      onParsedComplexBegin([]);
    },
    onArrayEnd: (offset, length, startLine, startCharacter) => {
      currentParent.length = offset + length - currentParent.offset;
      if (currentParent.range) {
        // @ts-ignore, read only ;P
        currentParent.range.end.line = startLine;
        // @ts-ignore, read only ;P
        currentParent.range.end.character = startCharacter + length;
      }
      currentParent = currentParent.parent!;
      ensurePropertyComplete(offset + length);

      onParsedComplexEnd();
    },
    onLiteralValue: (value, offset, length, startLine, startCharacter) => {
      onValue({
        type: getLiteralNodeType(value),
        offset,
        length,
        parent: currentParent,
        value,
        range: calculateRange(startLine, startCharacter, length),
      });
      ensurePropertyComplete(offset + length);

      onParsedValue(value);
    },
    onSeparator: (sep: string, offset: number, length: number) => {
      if (currentParent.type === 'property') {
        if (sep === ':') {
          currentParent.colonOffset = offset;
        } else if (sep === ',') {
          ensurePropertyComplete(offset);
        }
      }
    },
    onError: (error: ParseErrorCode, offset, length, startLine, startCharacter) => {
      errors.push({
        range: calculateRange(startLine, startCharacter, length),
        message: printParseErrorCode(error),
        severity: DiagnosticSeverity.Error,
        code: error,
      });
    },
  };
  visit(text, visitor, options);

  const result = currentParent.children![0];
  if (result) {
    delete result.parent;
  }
  return {
    ast: result,
    data: currentParsedParent[0],
    lineMap,
  };
}

function getLiteralNodeType(value: any): NodeType {
  switch (typeof value) {
    case 'boolean':
      return 'boolean';
    case 'number':
      return 'number';
    case 'string':
      return 'string';
    default:
      return 'null';
  }
}

// builds up the line map, for use by linesForPosition
const computeLineMap = (input: string) => {
  const lines = input.split(/\n/);
  const lineMap = [0];

  let sum = 0;
  for (const line of lines) {
    sum += line.length + 1;
    lineMap.push(sum);
  }

  return lineMap;
};
