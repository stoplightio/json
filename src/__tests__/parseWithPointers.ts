import * as fs from 'fs';
import { join } from 'path';
import { parseWithPointers } from '../parseWithPointers';

describe('json parser', () => {
  test('parse simple', () => {
    expect(
      parseWithPointers(`{
  "hello": "world",
  "address": {
    "street": 123
  },
  "paths": {
    "/users/{id}": {
      "get": {
        "operationId": "get-user"
      }
    }
  }
}`)
    ).toMatchSnapshot();
  });

  test('parse complex', () => {
    expect(
      parseWithPointers(`{
  "users": [
    {
      "name": "Markku",
      "address": {
         "country": "Finland",
         "city": "Helsinki"
       },
      "age": 17,
      "adult": false
    },
    {
      "name": "Dennis",
      "country": "Denmark",
      "age": 18,
      "adult": true
    }
  ]
}`)
    ).toMatchSnapshot();
  });

  test('does not allow comments by default', () => {
    expect(
      parseWithPointers(`{
      // a comment
      "name": "Antti",
      /* a
       very
       block 
       comment  
       */
      "city": "Vantaa"
    }`)
    ).toHaveProperty('validations', [
      expect.objectContaining({
        name: 'InvalidCommentToken',
        location: expect.objectContaining({
          start: expect.objectContaining({
            line: 2,
          }),
        }),
      }),
      expect.objectContaining({
        name: 'InvalidCommentToken',
        location: expect.objectContaining({
          start: expect.objectContaining({
            line: 4,
          }),
        }),
      }),
    ]);
  });

  test('does not allow trailing commas by default', () => {
    expect(
      parseWithPointers(`{
      "name": "Antti",
       "city": "Vantaa",
    }`)
    ).toHaveProperty('validations', [
      expect.objectContaining({
        name: 'PropertyNameExpected',
        location: {
          start: expect.objectContaining({
            line: 4,
          }),
          end: expect.objectContaining({
            line: 4,
          }),
        },
      }),
      expect.objectContaining({
        name: 'ValueExpected',
        location: {
          start: expect.objectContaining({
            line: 4,
          }),
          end: expect.objectContaining({
            line: 4,
          }),
        },
      }),
    ]);
  });

  describe('fixtures', () => {
    it.each(['main.oas2.json', 'user.jschema.json'])('parses %s', async filename => {
      expect(
        parseWithPointers((await fs.promises.readFile(join(__dirname, 'fixtures', filename), 'utf-8')) as string)
      ).toMatchSnapshot();
    });
  });

  describe('invalid fixtures', () => {
    it.each(['schema.json', 'characters.json'])('parses %s', async filename => {
      expect(
        parseWithPointers((await fs.promises.readFile(
          join(__dirname, 'fixtures/invalid', filename),
          'utf-8'
        )) as string)
      ).toMatchSnapshot();
    });
  });
});
