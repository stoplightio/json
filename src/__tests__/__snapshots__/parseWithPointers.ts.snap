// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`json parser fixtures parses petstore.oas2.json 1`] = `
Object {
  "ast": Any<Object>,
  "data": Object {
    "basePath": "/v2",
    "definitions": Object {
      "Category": Object {
        "properties": Object {
          "name": Object {
            "type": "string",
          },
        },
        "required": Array [
          "name",
        ],
        "title": "Category",
        "type": "object",
      },
      "Error": Object {
        "properties": Object {
          "code": Object {
            "type": "string",
          },
          "message": Object {
            "type": "string",
          },
        },
        "required": Array [
          "code",
        ],
        "type": "object",
      },
      "Pet": Object {
        "properties": Object {
          "category": Object {
            "$ref": "#/definitions/Category",
          },
          "id": Object {
            "format": "int64",
            "type": "integer",
          },
          "name": Object {
            "example": "doggie",
            "type": "string",
          },
          "photoUrls": Object {
            "items": Object {
              "type": "string",
            },
            "type": "array",
            "xml": Object {
              "name": "photoUrl",
              "wrapped": true,
            },
          },
          "status": Object {
            "description": "pet status in the store",
            "enum": Array [
              "available",
              "pending",
              "sold",
            ],
            "type": "string",
          },
        },
        "required": Array [
          "name",
          "photoUrls",
        ],
        "type": "object",
        "xml": Object {
          "name": "Pet",
        },
      },
    },
    "externalDocs": Object {
      "description": "Find out more about Swagger",
      "url": "http://swagger.io",
    },
    "host": "petstore.swagger.io",
    "info": Object {
      "contact": Object {
        "email": "apiteam@swagger.io",
      },
      "description": "This is a sample server Petstore server.  You can find out more about Swagger at [http://swagger.io](http://swagger.io) or on [irc.freenode.net, #swagger](http://swagger.io/irc/).  For this sample, you can use the api key \`special-key\` to test the authorization filters.",
      "license": Object {
        "name": "Apache 2.0",
        "url": "http://www.apache.org/licenses/LICENSE-2.0.html",
      },
      "termsOfService": "http://swagger.io/terms/",
      "title": "Swagger Petstore",
      "version": "1.0.0",
    },
    "parameters": Object {
      "limit": Object {
        "in": "query",
        "name": "limit",
        "type": "string",
      },
      "pet-id": Object {
        "in": "path",
        "name": "petId",
        "required": true,
        "type": "string",
      },
      "rate-limit": Object {
        "in": "header",
        "name": "Rate-Limit",
        "type": "string",
      },
      "skip": Object {
        "in": "query",
        "name": "skip",
        "type": "string",
      },
    },
    "paths": Object {
      "/pet/{petId}": Object {
        "parameters": Array [
          Object {
            "$ref": "#/parameters/pet-id",
          },
        ],
        "put": Object {
          "consumes": Array [
            "application/json",
            "application/xml",
          ],
          "description": "",
          "operationId": "updatePet",
          "parameters": Array [
            Object {
              "description": "Pet object that needs to be added to the store",
              "in": "body",
              "name": "body",
              "required": true,
              "schema": Object {
                "$ref": "#/definitions/Pet",
              },
            },
          ],
          "produces": Array [
            "application/xml",
            "application/json",
          ],
          "responses": Object {
            "400": Object {
              "description": "Invalid ID supplied",
            },
            "404": Object {
              "$ref": "#/responses/404",
            },
            "405": Object {
              "description": "Validation exception",
            },
          },
          "security": Array [
            Object {
              "petstore_auth": Array [
                "write:pets",
                "read:pets",
              ],
            },
          ],
          "summary": "Update an existing pet",
          "tags": Array [
            "pet",
          ],
        },
      },
      "/pets": Object {
        "get": Object {
          "parameters": Array [
            Object {
              "$ref": "#/parameters/skip",
            },
            Object {
              "$ref": "#/parameters/limit",
            },
            Object {
              "$ref": "#/parameters/rate-limit",
            },
          ],
          "responses": Object {
            "200": Object {
              "description": "",
              "schema": Object {
                "properties": Object {},
                "type": "object",
              },
            },
          },
          "summary": "List pets",
        },
        "post": Object {
          "consumes": Array [
            "application/json",
            "application/xml",
          ],
          "description": "",
          "operationId": "addPet",
          "parameters": Array [
            Object {
              "description": "Pet object that needs to be added to the store",
              "in": "body",
              "name": "body",
              "required": true,
              "schema": Object {
                "$ref": "#/definitions/Pet",
              },
            },
          ],
          "produces": Array [
            "application/xml",
            "application/json",
          ],
          "responses": Object {
            "405": Object {
              "description": "Invalid input",
            },
          },
          "security": Array [
            Object {
              "petstore_auth": Array [
                "write:pets",
                "read:pets",
              ],
            },
          ],
          "summary": "Add a new pet to the store",
          "tags": Array [
            "pet",
          ],
        },
      },
    },
    "responses": Object {
      "404": Object {
        "description": "Our shared 404 response.",
        "schema": Object {
          "$ref": "#/definitions/Error",
        },
      },
    },
    "schemes": Array [
      "https",
      "http",
    ],
    "securityDefinitions": Object {
      "api_key": Object {
        "in": "header",
        "name": "api_key",
        "type": "apiKey",
      },
      "petstore_auth": Object {
        "authorizationUrl": "https://petstore.swagger.io/oauth/dialog",
        "flow": "implicit",
        "scopes": Object {
          "read:pets": "read your pets",
          "write:pets": "modify pets in your account",
        },
        "type": "oauth2",
      },
    },
    "swagger": "2.0",
    "tags": Array [
      Object {
        "description": "Everything about your Pets",
        "externalDocs": Object {
          "description": "Find out more",
          "url": "http://swagger.io",
        },
        "name": "pet",
      },
      Object {
        "description": "Access to Petstore orders",
        "name": "store",
      },
      Object {
        "description": "Operations about user",
        "externalDocs": Object {
          "description": "Find out more about our store",
          "url": "http://swagger.io",
        },
        "name": "user",
      },
    ],
  },
  "diagnostics": Array [],
  "lineMap": Any<Array>,
}
`;

exports[`json parser fixtures parses user.jschema.json 1`] = `
Object {
  "ast": Any<Object>,
  "data": Object {
    "properties": Object {
      "age": Object {
        "maximum": 150,
        "minimum": 0,
        "type": "number",
      },
      "completed_at": Object {
        "format": "date-time",
        "type": "string",
      },
      "email": Object {
        "format": "email",
        "minLength": 2,
        "type": "string",
      },
      "items": Object {
        "items": Object {
          "type": Array [
            "string",
            "number",
          ],
        },
        "type": Array [
          "null",
          "array",
        ],
      },
      "name": Object {
        "description": "The user's full name.",
        "type": "string",
      },
      "permissions": Object {
        "properties": Object {
          "ids": Object {
            "items": Object {
              "type": "integer",
            },
            "type": "array",
          },
        },
        "type": Array [
          "string",
          "object",
        ],
      },
      "plan": Object {
        "anyOf": Array [
          Object {
            "properties": Object {
              "bar": Object {
                "type": "string",
              },
              "foo": Object {
                "type": "string",
              },
            },
            "required": Array [
              "foo",
              "bar",
            ],
            "type": "object",
          },
          Object {
            "items": Object {
              "type": "integer",
            },
            "type": "array",
          },
        ],
      },
      "ref": Object {
        "$ref": "#/definitions/error-response",
      },
    },
    "required": Array [
      "name",
      "age",
      "completed_at",
    ],
    "title": "User",
    "type": "object",
  },
  "diagnostics": Array [],
  "lineMap": Any<Array>,
}
`;

exports[`json parser invalid fixtures parses characters.json 1`] = `
Object {
  "ast": Any<Object>,
  "data": Object {
    "valid-escape": "\\\\n",
    "valid-unicode": "Å¼",
  },
  "diagnostics": Array [
    Object {
      "code": 1,
      "message": "InvalidSymbol",
      "range": Object {
        "end": Object {
          "character": 19,
          "line": 3,
        },
        "start": Object {
          "character": 13,
          "line": 3,
        },
      },
      "severity": 0,
    },
    Object {
      "code": 12,
      "message": "UnexpectedEndOfString",
      "range": Object {
        "end": Object {
          "character": 21,
          "line": 3,
        },
        "start": Object {
          "character": 19,
          "line": 3,
        },
      },
      "severity": 0,
    },
    Object {
      "code": 5,
      "message": "ColonExpected",
      "range": Object {
        "end": Object {
          "character": 21,
          "line": 3,
        },
        "start": Object {
          "character": 19,
          "line": 3,
        },
      },
      "severity": 0,
    },
    Object {
      "code": 15,
      "message": "InvalidEscapeCharacter",
      "range": Object {
        "end": Object {
          "character": 24,
          "line": 4,
        },
        "start": Object {
          "character": 20,
          "line": 4,
        },
      },
      "severity": 0,
    },
  ],
  "lineMap": Any<Array>,
}
`;

exports[`json parser invalid fixtures parses schema.json 1`] = `
Object {
  "ast": Any<Object>,
  "data": Object {
    "address": Object {},
    "hello": "world",
    "paths": Object {
      "/users/{id}": Object {
        "get": Object {
          "operationId": "get-user",
        },
      },
    },
  },
  "diagnostics": Array [
    Object {
      "code": 4,
      "message": "ValueExpected",
      "range": Object {
        "end": Object {
          "character": 3,
          "line": 4,
        },
        "start": Object {
          "character": 2,
          "line": 4,
        },
      },
      "severity": 0,
    },
    Object {
      "code": 6,
      "message": "CommaExpected",
      "range": Object {
        "end": Object {
          "character": 9,
          "line": 5,
        },
        "start": Object {
          "character": 2,
          "line": 5,
        },
      },
      "severity": 0,
    },
  ],
  "lineMap": Any<Array>,
}
`;

exports[`json parser parse complex 1`] = `
Object {
  "ast": Any<Object>,
  "data": Object {
    "users": Array [
      Object {
        "address": Object {
          "city": "Helsinki",
          "country": "Finland",
        },
        "adult": false,
        "age": 17,
        "name": "Markku",
      },
      Object {
        "adult": true,
        "age": 18,
        "country": "Denmark",
        "name": "Dennis",
      },
    ],
  },
  "diagnostics": Array [],
  "lineMap": Any<Array>,
}
`;

exports[`json parser parse simple 1`] = `
Object {
  "ast": Any<Object>,
  "data": Object {
    "address": Object {
      "street": 123,
    },
    "hello": "world",
    "paths": Object {
      "/users/{id}": Object {
        "get": Object {
          "operationId": "get-user",
        },
      },
    },
  },
  "diagnostics": Array [],
  "lineMap": Any<Array>,
}
`;
